### ISSUE TITLE
Live Example fails in Chrome on Moto X (2014)

### ISSUE THREAD

**Issue by john-parton** [serratus/quaggaJS#27]
Trying to follow the example with get user media on chrome for android. Preview looks like this:

I have a Moto X. Works fine on Firefox for android. Tried Chrome Beta and it had the same problem.

![screenshot_2015-04-15-10-13-32](https://cloud.githubusercontent.com/assets/2071543/7162148/a0bb2462-e358-11e4-8fc9-d78f418eeb7e.png)

**Comment by serratus** [comment]
That's odd. Currently I don't have access to a Moto X, but I just tried it on my LG G3 and it works just fine in Chrome Beta (42.0.2311.107). Which version of Chrome do you have?

![screenshot_2015-04-16-23-18-08](https://cloud.githubusercontent.com/assets/1381822/7191611/3ac5299a-e48f-11e4-8d80-8348c1fd4aba.png)

**Comment by john-parton** [comment]
Problem exists in both 41.0.2272.96 and 42.0.2311.107

Some totally off the wall guesses: is that there is some kind of mis-match between the resolution of the returned data versus the reported camera resolution? Maybe some kind of re-sampling going on? Or maybe the pixel data isn't being returned in RGB? Perhaps an endianness issue?

**Comment by serratus** [comment]
I was able to reproduce this phenomenon on my friends Moto X as well. That's really odd, since both phones (LG G3 and Moto X) have the same architecture, CPU and OS. I'll try to come up with a basic camera test demo in order to get to the root of this issue.

Right now, I have no idea why it happens. Have you tried to run it in Firefox for Android?

**Comment by john-parton** [comment]
I did try Firefox (said so in my first post) and it worked fine.

Nexus 4 works. First gen Moto X (the smaller one) works too.

I thought maybe it could be some customization that Motorola did on Android, but that seems unlikely.

My best guess is it's something specific to the second gen Moto X's camera.

If I get some time, I might look into it. Any guesses as to where I should start?

**Comment by serratus** [comment]
Alright, so this seems to be a device specific problem. I would say we start by creating a simple ``getUserMedia`` example which reads, manipulates and then displays the video again. I already have some code snippets that could help. But I can't address this issue before tomorrow. I'll let you know when I start working on it.

**Comment by serratus** [comment]
Let's give this [gist](https://gist.github.com/serratus/71eeac7b216dd584186a) a try. It is supposed to show a ``<video>`` above a ``<canvas>`` element of which both should show the current camera stream.

For now, this only works on Chrome, because of the constraints I've used (1280 x 720 px). You can also lower the resolution if you run into troubles. Due to security constraints, you have to serve the page from a web-server, and cannot run it directly from the filesystem.

**Comment by qgxspace** [comment]
ios system can support?

**Comment by serratus** [comment]
@qgxspace No, iOS is not supported because it lacks the ``getUserMedia`` API.

**Comment by john-parton** [comment]
This works pretty much exactly how you would expect. Doesn't have the "corruption"

**Comment by serratus** [comment]
Thanks for testing it on your Moto X. That's right, QuaggaJS converts the canvas-input to grayscale and then stores it in an ``Uint8Array``. There is a way to use the ``ImageWrapper`` but only in conjunction with the ``FrameGrabber`` which actually takes care of copying the data.

Right, the aspect ratio differences between the browsers. But this should not be an issue, because I'm initializing the buffers after the video is loaded reading the dimensions directly from the stream.

I'll try to come up with a simple demo which only uses the minimal count of components to actually make it representable. Stay tuned!

**Comment by stanver** [comment]
@serratus hello if I use codova to get barcode in ios,can i use quaggaJS?

**Comment by serratus** [comment]
@stanver As far as I know, that's currently not possible. There are no useful plugins available which make the camera stream available in the HTML context (no ``getUserMedia``). Maybe in the future. For now, the only way to go is by using the file-api for static images.

**Comment by 4thana** [comment]
the media capture plugin for cordova works also on iOS:
https://github.com/apache/cordova-plugin-media-capture

but of course it uses its own API instead of getUserMedia.

**Comment by serratus** [comment]
@4thana media-capture is a totally different story, since it does not allow to get access to the current stream that is being recorded. Only after the capturing is finished, the video/image can be used for further processing. That's the huge advantage of using ``getUserMedia``. Some people are experimenting with exposing the camera-stream to a canvas with cordova plugins (e.g.: [CanvasCamera](https://github.com/daraosn/Cordova-CanvasCamera))

**Comment by john-parton** [comment]
Maybe iOS support can be a separate issue instead of continuing to comment on this Moto X specific issue?

**Comment by serratus** [comment]
Sorry for the long absence, but I finally came up with a new demo for your problem. Can you give [this](https://dl.dropboxusercontent.com/u/7159706/quaggaJS/issues/27/index.html) a try and tell me if both, the ``<video>`` and ``<canvas>`` tag show the same image? What still puzzles me is the fact that the preview in the examples is a ``<video>`` element with the camera-stream attached. Nothing fancy, and no image processing whatsoever.

The link above simply demonstrates the first step of the quaggaJS pipeline, attaching the user's camera to the ``<video>`` element and copying that data to the ``<canvas>``. Have you tried to set ``numOfWorkers`` to ``0`` in the examples just to make sure that the workers don't interfere with the system?

**Comment by john-parton** [comment]
Both the canvas and video show the same corrupted image.

Maybe it's a bug with Chrome + Camera firmware?

**Comment by john-parton** [comment]
Haven't tried adjusting workers.

**Comment by john-parton** [comment]
Works perfectly in Firefox

**Comment by serratus** [comment]
I've just created a more simplified version which just uses the ``CameraAccess`` part and displays the current video-stream. Can you give [this](https://dl.dropboxusercontent.com/u/7159706/quaggaJS/issues/27/camera_test.html) a try?

**Comment by john-parton** [comment]
Still doesn't work.

![screenshot_2015-05-31-22-31-23](https://cloud.githubusercontent.com/assets/2071543/7905811/0cb958f4-07e5-11e5-988d-fb7b843bbcfa.png)

**Comment by xDarkShadowKnightx** [comment]
I'm having this same issue on my Moto X

**Comment by john-parton** [comment]
Just double checked the example you posted at https://github.com/serratus/quaggaJS/issues/27#issuecomment-107155439

Still fails on Android version 5.1 and Chrome 43.0.2357.93

**Comment by xDarkShadowKnightx** [comment]
I'm still getting the same issue. I'm also testing on Android 5.1 with Chrome 43.0.2357.93. I've actually decided to just write my application natively due to this! As this may be an issue on more than just the Moto X.

![screenshot_2015-07-28-10-18-00](https://cloud.githubusercontent.com/assets/4952718/8933701/38b51952-3512-11e5-97fd-d0affa3dec72.png)

**Comment by serratus** [comment]
Unfortunately I don't have any idea what's going on here. I guess the only way to find out is to report a bug directly at the Chromium team, since it seems to work on Firefox. I'll try to get my hands on a Moto X and create a minimal example producing this issue. Stay tuned!

@xDarkShadowKnightx Are you going to use zxing for your application? Using phonegap, or completely switching to native Android?

**Comment by john-parton** [comment]
Verified fixed on my device. Good work! :+1:
